nx/lib/nx/binary_backend.ex

```diff
@@ -2033,6 +2051,29 @@ defmodule Nx.BinaryBackend do
     from_binary(t, new_data)
   end
 
+  @impl true
+  def triangular_solve(
+        %{type: output_type} = b_holder,
+        %{type: input_type, shape: {rows, rows}} = a,
+        b
+      ) do
+    a_matrix = a |> to_binary() |> binary_to_matrix(input_type, {rows, rows})
+    b_vector = b |> to_binary() |> binary_to_matrix(input_type, {0, 1}) |> List.flatten()
+
+    x =
+      Enum.with_index(a_matrix)
+      |> Enum.zip(b_vector)
+      |> Enum.reduce([], fn {{row, idx}, current_b}, previous_y ->
+        if Enum.at(row, idx) == 0, do: raise(ArgumentError, "can't solve for singular matrix")
+
+        y = (current_b - dot_matrix(row, previous_y)) / Enum.at(row, idx)
+        previous_y ++ [y]
```

Comment:
I wonder if there is a way to implement this without the `++ [y]`, which can be expensive... 🤔

Reply:
I think I have an idea, which we can try out later. Pseudo code follows:

```elixir
a |> Enum.zip(b) |> Enum.reduce(%{row_idx: 0}, fn {row, current_b}, %{row_idx: row_idx} = y ->
  y = current_b - Enum.reduce_while(row, {0, 0}, fn 
    {x, col}, acc when col < row_idx -> {:cont, x * y[col] + acc}
    _, acc -> {:halt, acc}
end)
y |> Map.put(:row_idx, row_idx + 1) |> Map.put(row_idx, y)
end)
```

Reply:
The idea is to use maps for a constant-time `data[index]` access and update, so we can avoid needing to `Enum.reverse` either `previous_y` or `row`

Reply:
@polvalente Seems interesting. Is there a way to run memory benchmarks in Elixir or see how the allocations are going on behind the scenes?

Reply:
You could try Benchee or something similar